### PR TITLE
WIP: Added IndexRange for efficient region and grid space iteration

### PR DESCRIPTION
**Pull request by hjmjohnson**

### CODE REVIEW

**Comment by hjmjohnson** [comment]
http://review.source.kitware.com/#/c/23857/

Uploaded patch set 1.Nov 1 10:30 AM

Kitware Build Robot
Patch Set 1: Verified-1 Build Failed: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23857-1&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-01Nov 1 10:49 AM

Niels Dekker
Uploaded patch set 2.Nov 1 11:03 AM

Kitware Build Robot
Patch Set 2: Verified-1 Build Failed: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23857-2&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-01Nov 1 11:21 AM

Niels Dekker
Uploaded patch set 3.Nov 1 12:19 PM

Kitware Build Robot
Patch Set 3: Verified+1 Build Successful: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23857-3&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-01Nov 1 1:12 PM

Niels Dekker
Patch Set 3: The next patch should add support for image region iteration (starting at a possible non-zero index position, using itk::ImageRegion).5:10 AM

Niels Dekker
Uploaded patch set 4.9:13 AM

Kitware Build Robot
Patch Set 4: Verified-1 Build Failed: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23857-4&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-0110:23 AM

Niels Dekker
Patch Set 5: Patch Set 4 was rebased3:26 PM

Kitware Build Robot
Patch Set 5: Verified-1 Build Failed: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23857-5&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-014:19 PM

**Comment by N-Dekker** [comment]
@hjmjohnson Same question as for #105 : can you please explain why you closed this PR?

**Comment by hjmjohnson** [comment]
Accidently deleted/closed branch when the script for moving from Gerrit went haywire.

**Comment by thewtex** [comment]
Blowing my mind @N-Dekker ! :boom: Love it :heart:

File: Modules/Core/Common/include/itkIndexRange.h [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+  }
+
+
+  /** Constructs a range of indices for the specified image region.
+   * \note This function is unavailable when VBeginAtZero is true (in
+   * which case, it is removed by means of C++ SFINAE).
+   */
+  template <typename TDummy = typename std::enable_if<!VBeginAtZero, std::nullptr_t>::type>
+  explicit IndexRange(const ImageRegion<VDimension>& imageRegion, TDummy = nullptr)
+    :
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/Core/Common/include/itkIndexRange.h]: Will you be updating this to the correct way?

[the remark this answers is not in the document]
Comment (N-Dekker) [reply Modules/Core/Common/include/itkIndexRange.h]: @hjmjohnson I would certainly like to fix this SFINAE attempt, but I first have to figure out how to update a GitHub pull request from a fork of yours!

**Comment by hjmjohnson** [comment]
You can’t update mine.   Just fetch the PR locally,  and then make a new PR.

This is one advantage of gerrit.

**Comment by phcerdan** [comment]
You can open a merge request into @hjmjohnson branch in his fork, but way too convoluted! hehe

@N-Dekker This is great! Are you running any benchmark? It would be cool to have data about the improvements you are doing in Experimental!

**Comment by N-Dekker** [comment]
@phcerdan
> @N-Dekker This is great! Are you running any benchmark?

Thank, Pablo! So far I only did a benchmark to check if it's useful to distinguish _at compile-time_ between index iteration that starts at index [0, 0] (only parameterized by the image size, `itk::Size<N>`) and index iteration that may start anywhere (within a region specified by `itk::ImageRegion<N>{initialIndex, imageSize}`). I did not find much difference on forward iteration, but reverse iteration appears significantly faster (~10% on VS2017) when you can specify at compile time that the iteration range is only parameterized by the image size. So that appears a justification for the Boolean template parameter `VBeginAtZero`.

**Comment by N-Dekker** [comment]
> Blowing my mind @N-Dekker ! 💥 Love it ❤️

@thewtex Sorry Matt, didn't want to blow your mind. I really wanted to keep it simple! However, I felt challenged to add support for index iteration within an arbitrary `itk::ImageRegion` (including a possibly non-zero initial index), while still preserving maximum performance for the (common) case when the initial index is just [0, 0]. (See also my comment to Pablo!)

**Comment by N-Dekker** [comment]
Please continue further review at the PR from my fork: :point_right: #155

**Comment by hjmjohnson** [comment]
Superceeded by #155

**Pull request closed without merging**